File: gui/src/lib/series.ts
/**
 * What the series panel draws and sends, as pure functions (WP-1016).
 *
 * A series is N separate refinements chained by a warm start, and the two things
 * a panel over it must not get wrong are both about honesty rather than layout.
 *
 * **The order is the series**, so every edit — reorder, remove, a typed
 * coordinate — produces the whole list and one `PUT /api/series` sends it. The
 * arithmetic of "move this one up" is here because it is the only part of that a
 * test can judge without a DOM.
 *
 * **A smooth curve is exactly what a poisoned chain produces** (WP-0505's
 * measured lesson), so `SEQUENTIAL_PATH_DEPENDENT` is the headline and a flagged
 * trajectory is drawn differently rather than annotated in a footnote: the
 * panel gives a flagged parameter the warning colour, dashes its line
 * (`trajectoryLegend`), and — when `direction="both"` ran — draws the backward
 * chain beside it,
 * because the disagreement *is* the evidence and describing it in words would be
 * asking the reader to take it on trust.
 */

export interface SeriesPattern {
  upload: string;
  filename: string;
  label: string;
  x: number | null;
  reader: string;
  /** the *effective* reader keywords this member was read with */
  reader_options: Record<string, string>;
  n_points: number;
  two_theta_range: [number, number];
  /** the *file* carried esds — a mixed series is fitted under two weightings */
  has_sigma: boolean;
}

export interface SeriesSettings {
  carry: string[];
  refit: string;
  direction: string;
  x_label: string;
}

export interface SeriesSetup {
  patterns: SeriesPattern[];
  n_patterns: number;
  settings: SeriesSettings;
  choices: { refit: string[]; direction: string[] };
  carry_help: string;
  defaults: SeriesSettings;
  protocol: { mode: string; plan: string | null; n_stages: number };
  has_x: boolean;
  sigma_mixed: boolean;
  has_result: boolean;
  running: boolean;
}

export interface Trajectory {
  path: string;
  x: number[];
  x_label: string;
  value: number[];
  stderr: Array<number | null>;
  labels: string[];
  path_dependent: boolean;
  discontinuous: boolean;
  /** the backward chain's values, aligned with `x`, when `direction="both"` ran */
  backward: number[] | null;
  /** the largest forward/backward difference in combined σ; `null` when neither
   *  chain estimated an esd, which is where the fence itself abstains */
  n_sigma: number | null;
}

export interface SeriesEntry {
  index: number;
  label: string;
  x: number | null;
  status: string;
  statistics: { rwp: number; gof: number } | null;
  n_iterations: number;
  reseeded: boolean;
  rwp_warm: number | null;
  /** which attempt of the escalation ladder produced these values (WP-1051);
   *  the first pattern of a chain is always `"cold"`, having nothing to warm
   *  from, so this is *not* the same question as `reseeded` */
  rung: string;
  /** every rung attempted on this pattern, in ladder order */
  rungs_tried: string[];
  node_id: string | null;
  tree_id: string | null;
  diagnostics: Array<{ level: string; code: string; message: string }>;
}

/** What the panel sends as one member of the list. */
export function asRequest(patterns: SeriesPattern[]): Array<Record<string, unknown>> {
  return patterns.map((p) => ({
    upload: p.upload, label: p.label,
    // `null` and `undefined` are the same answer here — no coordinate — and the
    // server reads a missing key as "index is the axis"
    ...(p.x === null ? {} : { x: p.x }),
    ...(p.reader_options && Object.keys(p.reader_options).length
      ? { reader_options: p.reader_options } : {}),
  }));
}

/**
 * Move one member by `delta`, clamped — the list unchanged if it cannot move.
 *
 * Returning the same array reference when nothing moves is what lets a caller
 * skip the round trip: reordering the first item upward is a no-op, and a PUT
 * for it would rewrite the server's list to what it already holds and re-read
 * every file to do it.
 */
export function moveBy(patterns: SeriesPattern[], index: number,
                       delta: number): SeriesPattern[] {
  const to = index + delta;
  if (index < 0 || index >= patterns.length || to < 0 || to >= patterns.length) {
    return patterns;
  }
  const next = [...patterns];
  const [moved] = next.splice(index, 1);
  next.splice(to, 0, moved);
  return next;
}

/** Sort the whole list by its coordinate — the ramp a user meant, not the order
 *  a file picker happened to hand over.  Members with no coordinate keep their
 *  places at the end, because inventing one is what `x` being `null` refuses. */
export function sortByX(patterns: SeriesPattern[]): SeriesPattern[] {
  const withX = patterns.filter((p) => p.x !== null);
  const without = patterns.filter((p) => p.x === null);
  return [...withX.sort((a, b) => (a.x as number) - (b.x as number)), ...without];
}

/**
 * The trajectories worth showing first: the path-dependent ones, then the rest.
 *
 * Ranked on `n_sigma` **inside the flagged group only**, which is why the server
 * computes it — a `Diagnostic` carries `where` and no magnitude, so "which
 * parameter disagrees *most*" is not answerable from the fences alone (WP-1012 hit
 * the same wall; the answer was to compute it from the two chains rather than to
 * grow the schema).
 *
 * Sorting the *unflagged* ones by σ too was the first version and a browser
 * showed why not: on a clean ramp nothing is flagged, every distance is under
 * 5e-4 σ, and ordering 15 parameters by that noise put `phases.0.cell.a` — the
 * one anybody opens a ramp to look at — eighth, in an order that reads as random.
 * Below the fence there is nothing to rank, so the series' own first-seen order
 * stands (`Array.prototype.sort` is stable).
 */
export function rankTrajectories(trajectories: Trajectory[]): Trajectory[] {
  const score = (t: Trajectory) => (t.path_dependent ? 2 : t.discontinuous ? 1 : 0);
  return [...trajectories].sort((a, b) => {
    const diff = score(b) - score(a);
    if (diff !== 0) return diff;
    if (!a.path_dependent) return 0;
    return (b.n_sigma ?? 0) - (a.n_sigma ?? 0);
  });
}

/**
 * One trajectory's y-axis title: the parameter's **leaf**, not its path.
 *
 * Measured in a browser, and the reason it is not a character-count threshold
 * like `viz/plots.py:plot_trajectory`'s: a *rotated* title competes with the tick
 * labels for the same fixed left margin, so `phases.0.cell.a` — fifteen
 * characters, well under matplotlib's cut — rendered as `aes.0.cell.a` beside
 * `4.161`-style ticks. The width that clips is the plot's margin, not the string,
 * and the leaf is what a y-axis wants anyway; the full path is the heading
 * directly above.
 *
 * A derived curve keeps its metric on the axis rather than stripping to the
 * leaf: `qpa.LaB6` is a weight percentage, and `r_bragg.`/`r_f.` are the two
 * McCusker residuals (fractions, not percentages). Their leaf alone is the bare
 * phase name — `LaB6` — which carries no unit, reads as a parameter named after
 * the phase, and is identical between the two residuals; naming the metric is
 * what keeps R_Bragg and R_F apart and off a plain leaf.
 */
export function axisTitle(traj: Trajectory | null): string {
  if (!traj) return "";
  if (traj.path.startsWith("qpa.")) return `${traj.path.slice(4)} (wt %)`;
  if (traj.path.startsWith("r_bragg.")) return `${traj.path.slice(8)} R_Bragg`;
  if (traj.path.startsWith("r_f.")) return `${traj.path.slice(4)} R_F`;
  return traj.path.split(".").at(-1) ?? traj.path;
}

/** A one-line reading of a trajectory's standing — the panel's caption. */
export function trajectoryNote(traj: Trajectory | null, sigmaBar: number): string {
  if (!traj) return "";
  if (traj.path_dependent) {
    const n = traj.n_sigma === null ? "" : ` by up to ${traj.n_sigma.toFixed(1)}σ`;
    return `the forward and backward chains disagree${n} (over ${sigmaBar}σ): this `
      + "trajectory depends on the order the series was refined in, so it is not "
      + "determined by the data alone — hold the parameter, restrain it, or quote "
      + "the between-chain spread as its uncertainty";
  }
  if (traj.discontinuous) {
    return "a step far larger than this parameter's own scatter: either the "
      + "specimen changed here or the chain failed and carried the error onward "
      + "— open that pattern's own fit before reading the jump as physics";
  }
  if (traj.backward !== null) {
    return "the forward and backward chains agree within their esds";
  }
  return "";
}

/** One legend entry: the mark it toggles, what it says, and the custom property
 *  its swatch is drawn in, so a theme switch restyles it by CSS alone. */
export interface LegendEntry {
  id: string;
  label: string;
  ink: string;
  mark: "line" | "dash" | "dot" | "ring" | "cross";
}

/**
 * What the trajectory's legend offers — the marks `rxplot.trajectory` draws.
 *
 * A flagged parameter's forward chain wears `--warn` and is dashed, and its
 * name says why: the disagreement between the chains is the evidence, so it is
 * drawn rather than described. The backward chain is offered only when
 * `direction="both"` ran.
 *
 * Reseeded points are ringed and never dropped: the fit is good, but its
 * starting values did not come from its neighbour, so it is not evidence that
 * the trajectory is continuous there (`SEQUENTIAL_RESEED`). A point no rung of
 * the ladder recovered is **crossed** instead (`SEQUENTIAL_UNRECOVERED`,
 * WP-1051). The two marks say opposite things: a ring is a good fit reached
 * from a different starting model, a cross a diverged fit whose value is not a
 * measurement. The crossed point is still plotted, because a gap reads as data
 * nobody collected.
 *
 * A point with no esd carries no whisker at all, which the figure owns. A
 * well-determined trajectory then shows no visible whisker, and that is the
 * data rather than a defect: on the synthetic ramp under `mccusker_default`,
 * σ(a) is 6.5e-6 Å against a 4.8e-3 Å axis over 189 px, so a 2σ whisker is
 * 0.5 px. Scaling it up to be seen would be WP-1029's "an exaggeration is not
 * a probability" one panel over.
 */
export function trajectoryLegend(traj: Trajectory, reseeded: boolean[] = [],
                                 unrecovered: boolean[] = []): LegendEntry[] {
  const flagged = traj.path_dependent;
  const tone = flagged ? "--warn" : "--plot-diff";
  const out: LegendEntry[] = [
    { id: "forward", label: flagged ? "forward (path-dependent)" : "forward",
      ink: tone, mark: flagged ? "dash" : "line" },
  ];
  if (traj.stderr.some((e) => typeof e === "number" && Number.isFinite(e))) {
    out.push({ id: "esd", label: "esd", ink: tone, mark: "line" });
  }
  if (traj.backward) {
    out.push({ id: "backward", label: "backward", ink: "--muted", mark: "dot" });
  }
  if (reseeded.some(Boolean)) {
    out.push({ id: "rings", label: "reseeded", ink: "--warn", mark: "ring" });
  }
  if (unrecovered.some(Boolean)) {
    out.push({ id: "crosses", label: "unrecovered", ink: "--warn", mark: "cross" });
  }
  return out;
}

/** `v` to `n` significant figures, with no trailing zeros. */
const sig = (v: number, n: number) => String(Number(v.toPrecision(n)));

/**
 * What the pointer on one point says: the pattern, its coordinate, and the
 * value, with its esd when the fit gave one. The value takes six significant
 * figures, as the hover did under plotly (`%{y:.6g}`), and the esd two, as an
 * esd is quoted. A backward point is named as the backward chain's, since
 * both chains share every coordinate.
 */
export function pointText(traj: Trajectory, i: number,
                          chain: "forward" | "backward"): string {
  const where = `${traj.labels[i] ?? i} · ${sig(traj.x[i], 6)}`;
  if (chain === "backward") return `${where}: ${sig(traj.backward![i], 6)} (backward)`;
  const e = traj.stderr[i];
  const esd = typeof e === "number" && Number.isFinite(e) ? ` ± ${sig(e, 2)}` : "";
  return `${where}: ${sig(traj.value[i], 6)}${esd}`;
}

/**
 * What a member's pattern chart offers: the fitted points, the masked ones
 * when the protocol masked any, the model, the background and the residual.
 * `rxplot.pattern`'s ids, over its own inks.
 */
export function memberLegend(masked: boolean): LegendEntry[] {
  return [
    { id: "obs", label: "obs", ink: "--plot-obs", mark: "dot" },
    ...(masked ? [{ id: "masked", label: "excluded", ink: "--muted", mark: "dot" } as LegendEntry] : []),
    { id: "calc", label: "calc", ink: "--plot-calc", mark: "line" },
    { id: "bkg", label: "background", ink: "--plot-bkg", mark: "dash" },
    { id: "diff", label: "Δ/σ", ink: "--plot-diff", mark: "line" },
  ];
}

/** Which series entries carry a flag, by trajectory position.
 *
 * A trajectory skips patterns where its path is absent (`SeriesResult.trajectory`
 * does not fill gaps), so the flags are matched on **label** rather than on
 * index — the two are the same list only when every pattern carried the path. */
function flagsFor(traj: Trajectory, entries: SeriesEntry[],
                  flagged: (e: SeriesEntry) => boolean): boolean[] {
  const marked = new Set(entries.filter(flagged).map((e) => e.label));
  return traj.labels.map((label) => marked.has(label));
}

/** Which entries the reseed fence refitted cold. */
export function reseededFlags(traj: Trajectory, entries: SeriesEntry[]): boolean[] {
  return flagsFor(traj, entries, (e) => e.reseeded);
}

/** Which entries no rung of the ladder recovered (WP-1051).
 *
 * Read off `status` rather than a flag of its own, exactly as the library's
 * diagnostic is: "diverged after the last rung" and "diverged" are the same
 * statement once the chain has run, and a second field could disagree with it. */
export function unrecoveredFlags(traj: Trajectory,
                                 entries: SeriesEntry[]): boolean[] {
  return flagsFor(traj, entries, (e) => e.status === "diverged");
}

File: src/rietx/gui/series.py
"""The GUI's sequential-series surface — the setup, and the answer (WP-1016).

A series is N *separate* refinements chained by a warm start
(:mod:`rietx.sequential`), not one joint residual, and this module is where
that shape meets a session that owns exactly one project.  Three decisions carry
it, and each is a refusal to grow something.

**A series lives beside the project, not inside it.**  ``ProjectDoc.patterns``
stays length 1 and ``Project.open`` still refuses more, so the series' patterns
are staged uploads (WP-1014) held in the session, its per-pattern histories are
**in memory**, and its answer is session-scoped like ``Refinement.result_``.
That is not a shortcut: an upload token dies with the session
(``UploadStore.close``), so a *persisted* series would need a document to name
its files, which is exactly the schema growth the v1.0 GUI plan told this WP to
record for WP-1003 rather than perform.  What the project *does* supply is the
protocol — mode, plan, 2θ limits and excluded regions all come from
``project.json``, so one protocol is applied to N specimens and the series cannot
disagree with the single-pattern fit beside it about what is being fitted.

**The models come from the working state, and the settings from here.**  The
starting structure and instrument are the project's head — a series warm-starts
from whatever the user has already got right — while ``carry``/``refit``/
``direction``/``x_label`` are properties of the chain and have nowhere else to
live.  ``carry`` is a *control*, not a tuning knob (WP-0505 measured that
carrying everything costs 838 iterations against 904 for a narrower glob, with
identical Rwp), and :data:`CARRY_HELP` is that sentence where the editor can
show it.

**The disagreement between the two chains is reported as a number here, not as a
field on a diagnostic.**  ``SEQUENTIAL_PATH_DEPENDENT`` is a
:class:`~rietx.schemas.common.Diagnostic`, which carries ``where`` and no
magnitude — the wall WP-1012 hit displaying guard diagnostics per node, and left
to WP-1003 as a freeze question.  It does not need answering: the two chains'
trajectories are both in hand, so :func:`trajectories` recomputes the combined-σ
distance the same way :func:`sequential._path_dependence_diagnostics` does and
serves it per path.  A panel can then rank by disagreement without a schema
change.
"""

from __future__ import annotations

from dataclasses import dataclass, field, replace
from pathlib import Path
from typing import Any

import numpy as np

from ..schemas.pattern import PatternData
from ..schemas.sequential import SeriesResult
from ..sequential import (
    DIRECTIONS,
    PATH_DEPENDENCE_SIGMA,
    REFIT_MODES,
    _noise_floor,
    unique_labels,
)
from .imports import UploadStore, preview_pattern

#: What the carry-glob editor says about itself.  A control for parameters that
#: provably must not chain — never a knob to tune — and the measurement is why
#: (WP-0505, quoted in ``_carry_into``'s docstring).
CARRY_HELP = (
    "dot-path globs naming which parameters cross the pattern boundary; "
    "everything else restarts from the model this series began at. Default "
    "['*'] — carry everything. Measured on the round-robin sample-1 series: "
    "carrying everything costs 838 iterations against 904 for a carry that "
    "excludes the phase scales, with identical Rwp and identical weight "
    "fractions, so this is a control for a parameter that provably must not "
    "chain, not a knob to tune."
)

#: The setting defaults, in one place so the GET before any PUT and the run agree.
#: They are WP-0505's measured results and are deliberately not re-litigated in
#: the UI: ``refit="single"`` (904 iterations against 1623 staged, same answer)
#: and carry everything.
DEFAULTS: dict[str, Any] = {
    "carry": ["*"],
    "refit": "single",
    "direction": "forward",
    "x_label": "index",
}


class SeriesRefused(ValueError):
    """A setting or a member was refused, carrying the field at fault.

    The field matters as much as the sentence: a form highlights what to retype,
    and a refusal of ``refit`` that reports ``patterns`` sends the user to the
    wrong control — the same reason ``GuiSession._edit`` extracts the offending
    dot-path from a ``ParameterTable`` refusal (WP-1035).
    """

    def __init__(self, message: str, where: str) -> None:
        super().__init__(message)
        self.where = where


@dataclass
class SeriesMember:
    """One staged pattern in series order, with what reading it revealed."""

    token: str
    path: Path
    filename: str
    label: str
    #: the series coordinate — temperature, time, pressure.  ``None`` means the
    #: pattern index is the axis, which is what ``x_label`` then says.
    x: float | None
    reader: str
    #: the **effective** reader keywords this member was read with, as strings —
    #: ``DataRef.options`` one layer up, and the same vocabulary
    #: (:data:`~rietx.io.readers.READER_OPTIONS`), because a member of a series
    #: is a pattern file like any other and may need one to read as intended.
    options: dict[str, str]
    n_points: int
    two_theta_range: tuple[float, float]
    #: whether the *file* carried esds.  Kept per member because a series whose
    #: files disagree is a weighting inconsistency (CLAUDE.md, Weights) that is
    #: invisible once every pattern is read — see :func:`setup_payload`.
    has_sigma: bool

    def as_dict(self) -> dict:
        return {"upload": self.token, "filename": self.filename,
                "label": self.label, "x": self.x, "reader": self.reader,
                "reader_options": dict(self.options), "n_points": self.n_points,
                "two_theta_range": list(self.two_theta_range),
                "has_sigma": self.has_sigma}


@dataclass
class SeriesSetup:
    """The staged series and the chain's own settings — session-scoped."""

    members: list[SeriesMember] = field(default_factory=list)
    carry: list[str] = field(default_factory=lambda: list(DEFAULTS["carry"]))
    refit: str = DEFAULTS["refit"]
    direction: str = DEFAULTS["direction"]
    x_label: str = DEFAULTS["x_label"]

    @property
    def labels(self) -> list[str]:
        return [m.label for m in self.members]

    @property
    def x(self) -> list[float] | None:
        """The series coordinate, or ``None`` when any member lacks one.

        All or nothing on purpose: ``SeriesResult.x`` falls back to the *index*
        per entry, so a half-filled coordinate would put three temperatures and
        two indices on one axis and label it with whichever the caller named.
        """
        xs = [m.x for m in self.members]
        if not xs or any(v is None for v in xs):
            return None
        return [float(v) for v in xs]

    @property
    def axis_label(self) -> str:
        """What the trajectory axis is actually called.

        ``x_label`` is a *setting* and the coordinate may be absent, and
        ``SequentialRefinement.fit`` only renames the axis in the other
        direction (a given ``x`` promotes a default ``"index"`` to ``"x"``). So a
        user who typed ``T`` and then cleared one temperature would get an axis
        of pattern indices labelled ``T`` — a trajectory whose x values mean
        something other than what the label says, which is the one thing a
        plotted axis may never do.
        """
        return self.x_label if self.x is not None else "index"


# ----------------------------------------------------------------------
# the setup
# ----------------------------------------------------------------------
def members_from(entries: Any, uploads: UploadStore,
                 known: list[SeriesMember] | None = None) -> list[SeriesMember]:
    """Resolve ``[{upload, label?, x?}, …]`` against the session's staged files.

    Every member is *read* here, not at run time, which is WP-1014's two-phase
    property applied to N files: a file that does not parse is a message about
    that file, not a series that dies half way through a chain.  The description
    comes from :func:`preview_pattern` verbatim (reader, point count, range,
    ``has_sigma``), so the series list and the import wizard cannot disagree
    about what a file is.

    ``known`` is the list this one replaces, and a member already described in it
    is **not re-read**: a staged upload is immutable (its bytes are on disk under
    a token nothing rewrites), so the reader, the point count and ``has_sigma``
    cannot have changed.  It matters because the order *is* the series — every
    reorder, every typed coordinate is a whole-list PUT — and re-reading forty
    patterns per keystroke would make the editor unusable while proving nothing.
    """
    def _key(token: str, options: dict) -> tuple:
        return (token, tuple(sorted(options.items())))

    # keyed on the *effective* options, which is also what ``as_dict`` reports:
    # the panel sends back the list it was given, so a member it did not touch
    # keys identically and is not re-read.
    cached = {_key(m.token, m.options): m for m in (known or [])}
    if not isinstance(entries, list):
        raise SeriesRefused("'patterns' must be a list of "
                            "{upload: token, label?: str, x?: number}",
                            "patterns")
    out: list[SeriesMember] = []
    raw_labels: list[str] = []
    for i, entry in enumerate(entries):
        if isinstance(entry, str):
            entry = {"upload": entry}
        if not isinstance(entry, dict):
            raise SeriesRefused(f"patterns[{i}] must be an object with an "
                                "'upload' token", f"patterns.{i}")
        token = str(entry.get("upload") or "")
        if not token:
            raise SeriesRefused(f"patterns[{i}] has no 'upload' token",
                                f"patterns.{i}.upload")
        staged = uploads.get(token, "pattern")   # UploadRefused names the token
        # an empty string is a cleared control, not a request; anything else
        # goes through untouched so that a *typo* is refused in
        # ``reader_options_for``'s words rather than silently dropped here
        requested = {k: v for k, v in (entry.get("reader_options") or {}).items()
                     if v not in (None, "")}
        x = entry.get("x")
        if x is not None:
            try:
                x = float(x)
            except (TypeError, ValueError):
                raise SeriesRefused(f"patterns[{i}].x={x!r} is not a number",
                                    f"patterns.{i}.x") from None
        label = str(entry.get("label") or "").strip() or Path(staged.filename).stem
        raw_labels.append(label)
        seen = cached.get(_key(token, requested))
        if seen is None:
            preview = preview_pattern(staged, reader_options=requested)
            seen = SeriesMember(
                token=token, path=staged.path, filename=staged.filename,
                label=label, x=x, reader=preview["format"]["name"],
                options=dict(preview["reader_options"]),
                n_points=int(preview["n_points"]),
                two_theta_range=tuple(preview["two_theta_range"]),
                has_sigma=bool(preview["has_sigma"]))
        out.append(replace(seen, label=label, x=x))
    # the names the run will use, shown before it runs (``unique_labels``)
    for member, label in zip(out, unique_labels(raw_labels), strict=True):
        member.label = label
    return out


def check_settings(carry: Any, refit: Any, direction: Any) -> None:
    """Refuse a setting the chain would refuse, in the same words.

    ``REFIT_MODES``/``DIRECTIONS`` are :mod:`rietx.sequential`'s own tuples, so
    a menu offering a value the validator rejects is not expressible.
    """
    if not isinstance(carry, list) or not carry or not all(
            isinstance(g, str) and g for g in carry):
        raise SeriesRefused("'carry' is a non-empty list of dot-path globs; "
                            "['*'] carries everything", "carry")
    if refit not in REFIT_MODES:
        raise SeriesRefused(f"refit must be one of {REFIT_MODES}", "refit")
    if direction not in DIRECTIONS:
        raise SeriesRefused(f"direction must be one of {DIRECTIONS}", "direction")


def read_members(setup: SeriesSetup,
                 excluded_regions: list[tuple[float, float]]) -> list[PatternData]:
    """Read the staged files as the chain will fit them.

    The project's excluded regions are applied to **every** member: a series is
    one protocol over N specimens, and a region excluded because the sample
    holder scatters there is excluded in all of them.  The 2θ limits are the
    other half and travel as ``fit``'s own argument, since that is where the
    library takes them.
    """
    from ..io.readers import read_pattern

    out = []
    for member in setup.members:
        data = read_pattern(member.path, **member.options)
        data.excluded_regions = [tuple(r) for r in excluded_regions]
        out.append(data)
    return out


def setup_payload(setup: SeriesSetup, *, running: bool, has_result: bool,
                  mode: str, plan_preset: str | None, n_stages: int) -> dict:
    """The staged list, the settings, and the protocol the run will inherit.

    ``sigma_mixed`` is the one judgement here rather than a description: a series
    whose files disagree about carrying esds is fitted under two weighting
    policies, which is a correctness property invisible once the files are read
    (CLAUDE.md, Weights) and worth surfacing *before* the chain runs rather than
    as a footnote on its trajectories.
    """
    members = [m.as_dict() for m in setup.members]
    sigma = {m.has_sigma for m in setup.members}
    return {
        "patterns": members,
        "n_patterns": len(members),
        "settings": {"carry": list(setup.carry), "refit": setup.refit,
                     "direction": setup.direction, "x_label": setup.x_label},
        "choices": {"refit": list(REFIT_MODES), "direction": list(DIRECTIONS)},
        "carry_help": CARRY_HELP,
        "defaults": dict(DEFAULTS),
        # the protocol is the project's, quoted so a panel never re-derives it
        "protocol": {"mode": mode, "plan": plan_preset, "n_stages": n_stages},
        "has_x": setup.x is not None,
        "sigma_mixed": len(sigma) > 1,
        "has_result": has_result,
        "running": running,
    }


# ----------------------------------------------------------------------
# the answer
# ----------------------------------------------------------------------
def trajectories(series: SeriesResult,
                 backward: SeriesResult | None = None) -> list[dict]:
    """Every parameter's path across the series, plus the QPA fractions.

    One entry per dot-path in first-seen order, then one ``qpa.<phase>`` per
    phase a Rietveld fit weighed.  ``path_dependent`` and ``discontinuous`` are
    read off the series' **own** diagnostics rather than recomputed — the fences
    are the library's judgement — while ``n_sigma`` is computed here because no
    diagnostic carries a magnitude (see the module docstring).
    """
    unstable = {d.where[0] for d in series.diagnostics
                if d.code == "SEQUENTIAL_PATH_DEPENDENT" and d.where}
    jumps = {d.where[0] for d in series.diagnostics
             if d.code == "SEQUENTIAL_DISCONTINUITY" and d.where}
    names = list(series.paths(varied_only=False))
    phases = []
    for entry in series.entries:
        for row in (entry.qpa.phases if entry.qpa is not None else []):
            if row.name not in phases:
                phases.append(row.name)

    out = []
    for path in names + [f"qpa.{name}" for name in phases]:
        traj = (series.qpa_trajectory(path[4:]) if path.startswith("qpa.")
                else series.trajectory(path))
        if not len(traj):
            continue
        row = {"path": path, "x": list(traj.x), "x_label": traj.x_label,
               "value": list(traj.value), "stderr": list(traj.stderr),
               "labels": list(traj.labels),
               "path_dependent": path in unstable,
               "discontinuous": path in jumps,
               "backward": None, "n_sigma": None}
        if backward is not None and not path.startswith("qpa."):
            other = backward.trajectory(path)
            if len(other) == len(traj):
                row["backward"] = list(other.value)
                row["n_sigma"] = _disagreement(traj, other)
        out.append(row)
    return out


def _disagreement(forward, backward) -> float | None:
    """The largest forward/backward difference in combined σ, or ``None``.

    Deliberately the same arithmetic as
    :func:`sequential._path_dependence_diagnostics` — combined σ, the relative
    noise floor, ``nanargmax`` — so the number a panel sorts by is the number the
    fence fired on.  ``None`` where the fence itself abstains: a parameter with
    no esd in either chain cannot be judged this way, and reporting 0 would read
    as agreement it has not earned.
    """
    _, vf, sf = forward.arrays()
    _, vb, sb = backward.arrays()
    combined = np.sqrt(np.nan_to_num(sf) ** 2 + np.nan_to_num(sb) ** 2)
    if not np.any(combined > 0.0):
        return None
    with np.errstate(divide="ignore", invalid="ignore"):
        n_sigma = np.abs(vf - vb) / np.where(combined > 0.0, combined, np.nan)
    n_sigma = np.where(np.abs(vf - vb) > _noise_floor(vf, vb), n_sigma, 0.0)
    value = float(np.nanmax(n_sigma))
    return value if np.isfinite(value) else None


def result_payload(series: SeriesResult, backward: SeriesResult | None, *,
                   running: bool, curves: list[bool]) -> dict:
    """The series answer as a client needs it: entries, trajectories, fences.

    ``curves`` says per entry whether this session still holds that pattern's
    full :class:`~rietx.schemas.results.RefinementResult` — a
    :class:`SeriesResult` stores summaries, not curves (its module docstring), so
    the per-pattern plot is served from the runner in memory and a panel must
    know which rows it may open.

    ``path_dependent`` is hoisted to the top level because it is the **headline**
    and not a footnote: a smooth curve is exactly what a poisoned chain produces
    (WP-0505's measured lesson), so the one check that separates a measured
    trajectory from an ordering artefact cannot be something a user has to scroll
    to.
    """
    return {
        "result": series.model_dump(mode="json"),
        "trajectories": trajectories(series, backward),
        "path_dependent": sorted({d.where[0] for d in series.diagnostics
                                  if d.code == "SEQUENTIAL_PATH_DEPENDENT"
                                  and d.where}),
        "path_dependence_sigma": PATH_DEPENDENCE_SIGMA,
        "has_backward": backward is not None,
        "n_iterations": series.n_iterations,
        "curves": list(curves),
        "running": running,
    }
